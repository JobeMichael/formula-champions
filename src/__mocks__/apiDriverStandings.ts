export const apiDriverStandings = {
  MRData: {
    StandingsTable: {
      season: "2006",
      StandingsLists: [
        {
          season: "2006",
          DriverStandings: [
            {
              position: "1",
              points: "134",
              wins: "7",
              Driver: {
                driverId: "alonso",
                code: "ALO",
                givenName: "Fernando",
                familyName: "Alonso",
                nationality: "Spanish",
              },
              Constructors: [
                {
                  name: "Renault",
                },
              ],
            },
          ],
        },
      ],
    },
  },
};
