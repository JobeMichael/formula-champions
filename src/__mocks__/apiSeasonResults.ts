export const apiSeasonResults = {
  MRData: {
    xmlns: "http://ergast.com/mrd/1.4",
    series: "f1",
    url: "http://ergast.com/api/f1/2005/results/1.json",
    limit: "30",
    offset: "0",
    total: "19",
    RaceTable: {
      season: "2005",
      position: "1",
      Races: [
        {
          season: "2005",
          round: "1",
          url: "http://en.wikipedia.org/wiki/2005_Australian_Grand_Prix",
          raceName: "Australian Grand Prix",
          Circuit: {
            circuitId: "albert_park",
            url: "http://en.wikipedia.org/wiki/Melbourne_Grand_Prix_Circuit",
            circuitName: "Albert Park Grand Prix Circuit",
            Location: {
              lat: "-37.8497",
              long: "144.968",
              locality: "Melbourne",
              country: "Australia",
            },
          },
          date: "2005-03-06",
          time: "14:00:00Z",
          Results: [
            {
              number: "6",
              position: "1",
              positionText: "1",
              points: "10",
              Driver: {
                driverId: "fisichella",
                code: "FIS",
                url: "http://en.wikipedia.org/wiki/Giancarlo_Fisichella",
                givenName: "Giancarlo",
                familyName: "Fisichella",
                dateOfBirth: "1973-01-14",
                nationality: "Italian",
              },
              Constructor: {
                constructorId: "renault",
                url: "http://en.wikipedia.org/wiki/Renault_in_Formula_One",
                name: "Renault",
                nationality: "French",
              },
              grid: "1",
              laps: "57",
              status: "Finished",
              Time: {
                millis: "5057336",
                time: "1:24:17.336",
              },
              FastestLap: {
                rank: "2",
                lap: "55",
                Time: {
                  time: "1:25.994",
                },
              },
            },
          ],
        },
      ],
    },
  },
};
