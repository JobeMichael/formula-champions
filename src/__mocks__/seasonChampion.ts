export const seasonChampion = {
  season: "2006",
  points: "134",
  wins: "7",
  name: "Alonso Fernando",
  nationality: "Spanish",
  team: "Renault",
};
