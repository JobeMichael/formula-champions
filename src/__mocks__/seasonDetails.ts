export const seasonDetails = {
  date: "2005-03-06",
  name: "Fisichella Giancarlo",
  nationality: "Italian",
  raceName: "Australian Grand Prix",
  season: "2005",
  team: "Renault",
  time: "1:24:17.336",
};
