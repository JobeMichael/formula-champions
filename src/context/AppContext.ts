import React from "react";

export interface AppContextTypes {
  allChampions: Array<any>;
  setAllChampions: (currentData: any) => void;
}

export const URL_DEFAULT_VALUE = {
  allChampions: [],
  setAllChampions: () => {},
};

export const AppContext = React.createContext<AppContextTypes>(
  URL_DEFAULT_VALUE
);
