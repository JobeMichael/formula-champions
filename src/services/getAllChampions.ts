import axios from "axios";
import httpClient from "./httpClient";

interface Period {
  startYear: number;
  endYear: number;
}

interface GetAllChampions {
  (arg: Period): Promise<Array<ChampionsList>>;
}

export interface ChampionsList {
  name: string;
  nationality: string;
  points: string;
  season: string;
  team: string;
  wins: string;
  driverId: string;
}

const getAllChampions: GetAllChampions = async ({ startYear, endYear }) => {
  const instances = [];
  const { instance } = httpClient();

  while (endYear >= startYear) {
    instances.push(instance.get(`${startYear}/driverStandings.json`));
    startYear = startYear + 1;
  }

  const response = await axios
    .all(instances)
    .then(axios.spread((...responses) => responses))
    .catch((errors) => {
      console.log("Error!");
    });

  if (response) {
    return getFormattedSeasonsData(response);
  }
  return [];
};

const getFormattedSeasonsData = <T>(apiSeasonList: T[]): Array<ChampionsList> =>
  apiSeasonList.reduce((acc: any = [], item: any) => {
    const {
      season,
      StandingsLists: [
        {
          DriverStandings: [
            {
              points,
              wins,
              Driver: { familyName, givenName, nationality, driverId },
              Constructors: [{ name: team }],
            },
          ],
        },
      ],
    } = item.data.MRData.StandingsTable;

    acc.push({
      season,
      points,
      wins,
      name: `${familyName} ${givenName}`,
      nationality,
      team,
      driverId,
    });

    return acc;
  }, []);

export default getAllChampions;
