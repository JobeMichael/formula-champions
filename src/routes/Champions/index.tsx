import React, { useContext, useEffect, useState } from "react";
import { Spinner } from "react-bootstrap";
import ChampionList from "../../components/ChampionList";
import { AppContext } from "../../context/AppContext";
import getAllChampions from "../../services/getAllChampions";
import "./Champions.css";

const Champions = () => {
  const [loading, setLoading] = useState(false);
  const { setAllChampions, allChampions } = useContext(AppContext);

  useEffect(() => {
    const fetchSeasonsData = async () => {
      setLoading(true);
      const startYear = 2005;
      const endYear = 2015;
      const data = await getAllChampions({ startYear, endYear });

      setAllChampions((prevState: any) => [...prevState, ...data]);
      setLoading(false);
    };

    if (allChampions.length === 0) {
      fetchSeasonsData();
    }
  }, []);

  return (
    <div className="champions-wrapper">
      <div className="champions-header">
        <h1>2005 - 2015</h1>
        <p> Formula1 races winners from 2005 to 2015.</p>
      </div>
      {allChampions && <ChampionList data={allChampions} />}
      {loading && (
        <div className="loader-holder">
          <Spinner animation="border" variant="info" />
        </div>
      )}
    </div>
  );
};

export default Champions;
