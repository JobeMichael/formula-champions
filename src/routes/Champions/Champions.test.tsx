import { cleanup, render, waitForElement } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import Champions from ".";
import { AppContext } from "../../context/AppContext";
import getAllChampions from "../../services/getAllChampions";
import { seasonChampion } from "../../__mocks__/seasonChampion";

jest.mock("../../services/getAllChampions.ts");

describe("<Champions/>", () => {
  const setAllChampions = jest.fn();

  beforeEach(() => {
    setAllChampions.mockRestore();

    cleanup();
  });

  const setUp = (context) => (
    <AppContext.Provider value={context}>
      <MemoryRouter>
        <Champions />
      </MemoryRouter>
    </AppContext.Provider>
  );

  it("should get data from context and render component properly", async () => {
    const mockContext = {
      allChampions: [seasonChampion],
      setAllChampions: () => {},
    };
    const { asFragment, getByText } = render(setUp(mockContext));

    await waitForElement(() => getByText("Fernando Alonso"));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should get data from API", async () => {
    getAllChampions.mockReturnValue([seasonChampion]);
    const mockContext = {
      allChampions: [],
      setAllChampions: setAllChampions,
    };
    const { getByText } = render(setUp(mockContext));
    await waitForElement(() => getByText("Nationality"));

    expect(getAllChampions).toBeCalledTimes(1);
  });
});
