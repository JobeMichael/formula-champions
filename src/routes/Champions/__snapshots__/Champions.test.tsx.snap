// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Champions/> renders Champions correctly 1`] = `
<DocumentFragment>
  <div
    class="champions-wrapper"
  >
    <div
      class="champions-header"
    >
      <h1>
        2005 - 2015
      </h1>
      <p>
         Formula1 races winners from 2005 to 2015.
      </p>
    </div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Season
            </th>
            <th>
              Name
            </th>
            <th>
              Nationality
            </th>
            <th>
              Team
            </th>
            <th>
              Wins
            </th>
            <th>
              Points
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              2006
            </td>
            <td>
              Alonso Fernando
            </td>
            <td>
              Spanish
            </td>
            <td>
              Renault
            </td>
            <td>
              7
            </td>
            <td>
              134
            </td>
            <td
              class="button-holder"
            >
              <a
                href="season/2006"
              >
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Details
                </button>
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
